modules/c++/mem/include/mem/ScratchMemory.h

```diff
@@ -0,0 +1,140 @@
+/* =========================================================================
+ * This file is part of mem-c++
+ * =========================================================================
+ *
+ * (C) Copyright 2004 - 2018, MDA Information Systems LLC
+ *
+ * mem-c++ is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation; either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this program; If not,
+ * see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#ifndef __MEM_SCRATCHMEMORY_H__
+#define __MEM_SCRATCHMEMORY_H__
+
+#include <stddef.h>
+#include <map>
+#include <sstream>
+#include <string>
+#include <utility>
+#include <vector>
+#include <except/Exception.h>
+#include <mem/BufferView.h>
+#include <sys/Conf.h>
+
+namespace mem
+{
+
+/*!
+ *  \class ScratchMemory
+ *  \brief Handle reservation of scratch memory segments within a single buffer.
+ *
+ *  Memory segments should be reserved during a "setup" phase using the put
+ *  method. Once all segments are reserved, call the setup method to set up
+ *  the underlying memory and ensure the alignment requirements of each segment.
+ *  The get method may be used afterwards to obtain pointers to the memory
+ *  segments.
+ */
+class ScratchMemory
+{
+public:
+    ScratchMemory() :
+        mNumBytesNeeded(0),
+        mIsSetup(false)
+    {
+    }
+
+    /*!
+     * \brief Reserve a buffer segment within this scratch memory buffer.
+     *
+     * \param key Identifier for scratch segment
+     * \param numBytes Size of scratch buffer
+     * \param numBuffers Number of distinct buffers to set up
+     * \param alignment Number of bytes to align segment pointer
+     *
+     * \throws except::Exception if the given key has already been used
+     */
+    template <typename T>
+    void put(const std::string& key,
+             size_t numBytes,
+             size_t numBuffers = 1,
+             size_t alignment = sys::SSE_INSTRUCTION_ALIGNMENT);
+
+    /*!
+     * \brief Get pointer to buffer segment.
+     *
+     * \param key Identifier for scratch segment
+     * \param indexBuffer Index of distinct buffer
+     *
+     * \return Pointer to buffer segment
+     *
+     * \throws except::Exception if the scratch memory has not been set up,
+     *         the key does not exist, or index of buffer is out of bounds
+     */
+    template <typename T>
+    T* get(const std::string& key, size_t indexBuffer = 0);
```

Comment:
Is it ever useful to pass around a `const` `ScratchMemory` that can provide you `const T*`?  If it is, add a const overloading to `get()`.  If not, don't worry about it.

Reply:
I think there could be situations where ScratchMemory is a member of some object and get is called indirectly using a const reference to that object. I added a const version using the Effective C++ method (implement the const version, delegate to it from the non-const version and cast away the const). Let me know if that is alright with you.

Reply:
I reimplemented this as Clyde requested below:
https://github.com/mdaus/coda-oss/pull/267#discussion_r218024987